File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    truck_dq = deque(truck_weights)
    
    curr_bridge = deque([0] * bridge_length)
    curr_weight = 0
    while curr_bridge:
        answer += 1
        curr_weight -= curr_bridge.popleft()
        
        if truck_dq:
            if curr_weight + truck_dq[0] <= weight:
                curr_bridge.append(truck_dq.popleft())
                curr_weight += curr_bridge[-1]
            else: # 다리에 트럭 더 못 올라감
                curr_bridge.append(0)
        
    return answer

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        
        Queue<Integer> trucks = new LinkedList<>(
            Arrays.asList(Arrays.stream(truck_weights).boxed().toArray(Integer[]::new))
        ); // truck_weights를 Queue로 변환
        
        Queue<Integer> currBridge = new LinkedList<>(
            Collections.nCopies(bridge_length, 0)
        ); // bridge_length 만큼 0으로 초기화
        
        int currWeight = 0;
        
        while (!currBridge.isEmpty()) {
            answer++;
            currWeight -= currBridge.poll(); // 1초에 다리 한 칸 전진
            
            if (!trucks.isEmpty()) {
                int nextTruck = trucks.peek();
                if (currWeight + nextTruck <= weight) {
                    currBridge.offer(trucks.poll());
                    currWeight += nextTruck;
                } else {
                    currBridge.offer(0); // 다리에 트럭이 추가로 올라가지 못 함
                }
            }
        }
        
        return answer;
    }
}